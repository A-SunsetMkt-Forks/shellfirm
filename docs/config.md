# Config

When you install `shellfirm` the first time it creates a new config file in the home directory in the path: `~/.shellfirm/config.yaml`.
The main configuration file manage in [config.yaml](../src/config.yaml). 

You can always change your config file content and the `shellfirm` will never change it back. 
[read here how to add and test new command](./contributing.md#test-new-command)


# Config
| Parameter | Description | Values |
| --- | --- | --- |
| `challenge` | The way that you want to solve the challenge when risky command detected | `Math`, `Enter`, `Yes` |
| `includes` | List of group checks. | `list` |
| `checks[].is` | The value of the check | `String` |
| `checks[].method` | How to make the check | `Contains`, `Regex`, `StartWith` |
| `checks[].enable` | Enable disable the check | `true`, `false` |
| `checks[].description` | Prompt description when risky command directed | `String` |
| `checks[].from` | Group name | `String` |


# Update config file

Adding new groups:
```bash
$ shellfirm config update --check-group {group} {group}
```

Remove groups
```bash
$ shellfirm config update --check-group {group} {group} --remove
```

Reset 
```bash
$ shellfirm config reset
```