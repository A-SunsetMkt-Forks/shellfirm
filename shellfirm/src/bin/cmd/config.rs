use anyhow::{anyhow, Result};
use clap::{App, AppSettings::ArgRequiredElseHelp, Arg, ArgMatches, Command};
use shellfirm::{dialog, Challenge, Config, Settings};
use strum::IntoEnumIterator;

const ALL_GROUP_CHECKS: &[&str] = &include!(concat!(env!("OUT_DIR"), "/all_the_files.rs"));

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Manage app config")
        .setting(ArgRequiredElseHelp)
        .subcommand(
            App::new("update-groups")
                .about("enable check group")
                .arg(Arg::new("check-group").help("Check group")),
        )
        .subcommand(App::new("reset").about("Reset configuration"))
        .subcommand(App::new("challenge").about("Reset configuration"))
        .subcommand(App::new("ignore").about("Ignore command pattern"))
}

pub fn run(
    matches: &ArgMatches,
    config: &Config,
    settings: &Settings,
) -> Result<shellfirm::CmdExit> {
    match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("update-groups", _subcommand_matches) => {
                run_update_groups(config, &config.get_settings_from_file()?, None)
            }
            ("reset", _subcommand_matches) => Ok(run_reset(config, None)),
            ("challenge", _subcommand_matches) => run_challenge(config, None),
            ("ignore", _subcommand_matches) => run_ignore(config, settings, None),
            _ => unreachable!(),
        },
    }
}

pub fn run_update_groups(
    config: &Config,
    settings: &Settings,
    groups: Option<Vec<String>>,
) -> Result<shellfirm::CmdExit> {
    let check_groups = if let Some(groups) = groups {
        groups
    } else {
        let all_groups = ALL_GROUP_CHECKS.iter().map(|f| (*f).to_string()).collect();
        dialog::multi_choice(
            "select checks",
            all_groups,
            settings.get_active_groups().clone(),
            100,
        )?
    };

    match config.update_check_groups(check_groups) {
        Ok(()) => Ok(shellfirm::CmdExit {
            code: exitcode::OK,
            message: None,
        }),
        Err(e) => Ok(shellfirm::CmdExit {
            code: exitcode::CONFIG,
            message: Some(format!("Could not update checks group. error: {}", e)),
        }),
    }
}

pub fn run_reset(config: &Config, force_selection: Option<usize>) -> shellfirm::CmdExit {
    match config.reset_config(force_selection) {
        Ok(()) => shellfirm::CmdExit {
            code: exitcode::OK,
            message: Some("shellfirm configuration reset successfully".to_string()),
        },
        Err(e) => shellfirm::CmdExit {
            code: exitcode::CONFIG,
            message: Some(format!("reset settings error: {:?}", e)),
        },
    }
}

pub fn run_challenge(config: &Config, challenge: Option<Challenge>) -> Result<shellfirm::CmdExit> {
    let selection_challenge = if let Some(c) = challenge {
        c
    } else {
        let challenges = Challenge::iter().map(|c| c.to_string()).collect::<Vec<_>>();
        Challenge::from_string(&dialog::select("change shellfirm challenge", &challenges)?)?
    };

    match config.update_challenge(selection_challenge) {
        Ok(()) => Ok(shellfirm::CmdExit {
            code: exitcode::OK,
            message: None,
        }),
        Err(e) => Ok(shellfirm::CmdExit {
            code: exitcode::CONFIG,
            message: Some(format!("change challenge error: {:?}", e)),
        }),
    }
}

pub fn run_ignore(
    config: &Config,
    settings: &Settings,
    force_ignore: Option<Vec<String>>,
) -> Result<shellfirm::CmdExit> {
    let all_check_ids: Vec<String> = settings
        .get_active_checks()?
        .iter()
        .map(|c| c.id.to_string())
        .collect();

    let selected = if let Some(force_ignore) = force_ignore {
        force_ignore
    } else {
        dialog::multi_choice(
            "select checks",
            all_check_ids,
            settings.ignores.clone(),
            100,
        )?
    };

    match config.update_ignores(selected) {
        Ok(()) => Ok(shellfirm::CmdExit {
            code: exitcode::OK,
            message: None,
        }),
        Err(e) => Ok(shellfirm::CmdExit {
            code: exitcode::CONFIG,
            message: Some(format!("update pattern ignore errors: {:?}", e)),
        }),
    }
}

#[cfg(test)]
mod test_config_cli_command {

    use std::fs;

    use insta::{assert_debug_snapshot, with_settings};
    use tempdir::TempDir;

    use super::*;

    fn initialize_config_folder(temp_dir: &TempDir) -> Config {
        let temp_dir = temp_dir.path().join("app");
        Config::new(Some(&temp_dir.display().to_string())).unwrap()
    }

    #[test]
    fn can_run_update_groups() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        assert_debug_snapshot!(run_update_groups(
            &config,
            &config.get_settings_from_file().unwrap(),
            Some(vec!["test-1".to_string()])
        ));
        assert_debug_snapshot!(config.get_settings_from_file());
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_update_groups_with_error() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        let settings = config.get_settings_from_file().unwrap();
        fs::remove_file(&config.setting_file_path).unwrap();
        with_settings!({filters => vec![
            (r"error:.+", "error message"),
        ]}, {
            assert_debug_snapshot!(run_update_groups(
            &config,
            &settings,
            Some(vec!["test-1".to_string()])
        ));
        });
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_reset() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        config.update_challenge(Challenge::Yes).unwrap();
        assert_debug_snapshot!(run_reset(&config, Some(1)));
        assert_debug_snapshot!(config.get_settings_from_file());
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_reset_with_error() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        fs::remove_file(&config.setting_file_path).unwrap();
        with_settings!({filters => vec![
            (r"error:.+", "error message"),
        ]}, {
            assert_debug_snapshot!(run_reset(&config, Some(1)));
        });
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_challenge() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        let settings = config.get_settings_from_file().unwrap();
        assert_debug_snapshot!(run_challenge(&config, Some(Challenge::Yes)));
        assert_debug_snapshot!(
            config.get_settings_from_file().unwrap().challenge != settings.challenge
        );
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_challenge_with_err() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        fs::remove_file(&config.setting_file_path).unwrap();

        with_settings!({filters => vec![
            (r"error:.+", "error message"),
        ]}, {
            assert_debug_snapshot!(run_challenge(&config, Some(Challenge::Yes)));
        });
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_run_ignore() {
        let temp_dir = TempDir::new("config-app").unwrap();
        let config = initialize_config_folder(&temp_dir);
        let settings = config.get_settings_from_file().unwrap();
        assert_debug_snapshot!(run_ignore(
            &config,
            &settings,
            Some(vec!["id-1".to_string(), "id-2".to_string()])
        ));
        assert_debug_snapshot!(config.get_settings_from_file().unwrap().ignores);
        temp_dir.close().unwrap();
    }
}
