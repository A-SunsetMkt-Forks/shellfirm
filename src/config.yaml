challenge: Math # Math, Enter, YesNo

checks:
  - is: git reset
    method: Contains
    enable: true
    description: "This command going to reset all your local changes"
  - is: "rm.+(-r|-f|-rf|-fr)*"
    method: Regex
    enable: true
    description: "You are going to deletes everything in the path"
  - is: ":(){ :|: & };:"
    method: Contains
    enable: true
    description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack."
  - is: ">.+/dev/sda"
    method: Regex
    enable: true
    description: "Writing the data directly to the hard disk drive and damaging your file system."
  - is: "mv+.*/dev/null"
    method: Regex
    enable: true
    description: "The files will be discarded and destroyed."