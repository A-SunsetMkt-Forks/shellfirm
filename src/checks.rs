///! Manage command checks
///
use crate::config::{Challenge, Method};
use rand::Rng;
use rayon::prelude::*;
use regex::Regex;
use serde_derive::Deserialize;
use std::io;

/// Describe single check
#[derive(Debug, Deserialize, Clone)]
pub struct Check {
    pub is: String,
    pub method: Method,
    pub enable: bool,
    pub description: String,
}

impl Check {
    /// Show challenge to the user.
    pub fn show(&self, challenge: &Challenge) -> bool {
        match challenge {
            Challenge::Math => self.prompt_math(),
            Challenge::Enter => self.prompt_enter(),
            Challenge::YesNo => self.prompt_yesno(),
        }
    }

    /// Show prompt text + details of the challenge.
    ///
    /// # Arguments
    ///
    /// * `extra` - String with more text to the prompt question (usually for more detail of how solve the question).
    fn prompt_text(&self, extra: String) -> String {
        format!("{}, {}", self.description, extra)
    }

    /// Show math challenge prompt question to the user. creates random number between 0-10.
    fn prompt_math(&self) -> bool {
        let mut rng = rand::thread_rng();
        let num_a = rng.gen_range(0..10);
        let num_b = rng.gen_range(0..10);
        let expected_answer = num_a + num_b;

        eprintln!("{}", self.prompt_text(format!("{} + {} = ?", num_a, num_b)));
        loop {
            let answer = self.show_stdin_prompt();

            let answer: u32 = match answer.trim().parse() {
                Ok(num) => num,
                Err(_) => continue,
            };
            if answer == expected_answer {
                break;
            }
            eprintln!("wrong answer, try again...");
        }
        true
    }

    /// Show enter challenge to the user.
    fn prompt_enter(&self) -> bool {
        eprintln!("{}", self.prompt_text(format!("Type `Enter` to continue")));

        loop {
            let answer = self.show_stdin_prompt();
            if answer == "\n" {
                break;
            }
            eprintln!("wrong answer, try again...");
        }
        true
    }

    /// Show enter yes/no challenge to the user.
    fn prompt_yesno(&self) -> bool {
        eprintln!(
            "{}",
            self.prompt_text(format!("Type `yes` to continue `no` to cancel"))
        );
        let mut is_approve = true;

        loop {
            let answer = self.show_stdin_prompt();
            let answer = answer.trim();
            if answer == "yes" {
                break;
            } else if answer == "no" {
                is_approve = false;
                break;
            }
            eprintln!("wrong answer, try again...");
        }
        is_approve
    }

    /// Catch user stdin.
    fn show_stdin_prompt(&self) -> String {
        let mut answer = String::new();
        io::stdin()
            .read_line(&mut answer)
            .expect("Failed to read line");

        answer
    }
}

/// Check if the given command matched to on of the checks
///
/// # Arguments
///
/// * `checks` - List of checks that we want to validate.
/// * `command` - Command check.
pub fn run_check_on_command(checks: &[Check], command: &str) -> Vec<Check> {
    checks
        .par_iter()
        .filter(|&v| v.enable)
        .filter(|&v| is_match(v, command))
        .map(|v| v.clone())
        .collect()
}

/// returns true/false if the check match to the given command
fn is_match(check: &Check, command: &str) -> bool {
    match check.method {
        Method::Contains => is_contains(&check.is, command),
        Method::StartWith => is_start_with(&check.is, command),
        Method::Regex => is_regex(&check.is, command),
    }
}

// Checks if the given check contains the command.
fn is_contains(check: &str, command: &str) -> bool {
    command.contains(check)
}

// Checks if the given check start with the command.
fn is_start_with(check: &str, command: &str) -> bool {
    command.starts_with(check)
}

// Checks if the given check match to the command using regex.
fn is_regex(r: &str, command: &str) -> bool {
    // let search_regex = format!("r\"{}\"", r);
    Regex::new(r).unwrap().is_match(command)
}
