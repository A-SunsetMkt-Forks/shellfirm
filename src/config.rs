//! Configuration management

use crate::checks::Check;
use crate::cli::{UPDATE_CONFIGURATION_ONLY_DIFF, UPDATE_CONFIGURATION_OVERRIDE};
use anyhow::anyhow;
use anyhow::Result as AnyResult;
use serde_derive::{Deserialize, Serialize};
use std::fs;
use std::io::{Read, Write};

pub const DEFAULT_CONFIG_FILE: &str = include_str!("config.yaml");

/// The method type go the check.
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum Method {
    /// If the command start with.
    StartWith,
    /// if the command contains.
    Contains,
    /// if the command match to the given regex.
    Regex,
}

/// The user challenge when user need to confirm the command.
#[derive(Debug, Deserialize, Serialize)]
pub enum Challenge {
    /// Math challenge.
    Math,
    /// Only enter will approve the command.
    Enter,
    /// only yes/no typing will approve the command.
    YesNo,
}

/// describe configuration folder
pub struct SettingsConfig {
    /// Configuration folder path.
    pub path: String,
    /// config file.
    pub config_file_path: String,
    /// If configuration path overridden by the user.
    pub default: bool,
}

/// Describe the configuration yaml
#[derive(Debug, Deserialize, Serialize)]
pub struct Config {
    /// Type of the challenge.
    pub challenge: Challenge,
    /// List of checks.
    pub checks: Vec<Check>,
}

// /// Describe single check
// #[derive(Debug, Deserialize, Clone)]
// pub struct Check {
//     pub is: String,
//     pub method: Method,
//     pub enable: bool,
//     pub description: String,
// }

impl SettingsConfig {
    /// Convert config yaml file to struct.
    pub fn load_config_from_file(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(&self.read_config_file()?)?)
    }

    /// Return default app config as a config struct.
    pub fn load_default_config(&self) -> AnyResult<Config> {
        Ok(serde_yaml::from_str(DEFAULT_CONFIG_FILE)?)
    }

    /// Manage configuration folder/ file.
    /// * Create config folder if not exists
    /// * Create config yaml file if not exists
    pub fn manage_config_file(&self) -> AnyResult<()> {
        self.create_config_folder()?;
        if fs::metadata(&self.config_file_path).is_err() {
            self.create_default_config_file()?;
        }
        Ok(())
    }

    pub fn update_config_content(&self, behavior: &str) -> AnyResult<()> {
        match behavior {
            UPDATE_CONFIGURATION_ONLY_DIFF => self.add_diff_configuration_file(),
            UPDATE_CONFIGURATION_OVERRIDE => self.override_configuration_file(),
            _ => return Err(anyhow!("unsupported behavior")),
        }
    }

    /// Create config folder if not exists.
    fn create_config_folder(&self) -> AnyResult<()> {
        if let Err(err) = fs::create_dir(&self.path) {
            if err.kind() != std::io::ErrorKind::AlreadyExists {
                return Err(anyhow!("could not create folder: {}", err.to_string()));
            }
        }
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_default_config_file(&self) -> AnyResult<()> {
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(DEFAULT_CONFIG_FILE.as_bytes())?;
        Ok(())
    }

    /// Create config file with default configuration content
    fn create_config_file_from_struct(&self, config: &Config) -> AnyResult<()> {
        let content = serde_yaml::to_string(config)?;
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }

    /// Convert config file to config struct.
    fn read_config_file(&self) -> AnyResult<String> {
        let mut file = std::fs::File::open(&self.config_file_path)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        Ok(content)
    }

    /// override existing config yaml file with the application configuration
    fn override_configuration_file(&self) -> AnyResult<()> {
        self.create_default_config_file()
    }

    /// adding only new check into configuration file.
    fn add_diff_configuration_file(&self) -> AnyResult<()> {
        match self.load_config_from_file() {
            Ok(mut conf) => {

                let default_app_config =  self.load_default_config()?;

                for default_check in default_app_config.checks.clone() {
                    let mut found = false;
                    for user_check in &conf.checks{
                        if user_check.is == default_check.is {found = true; break;}
                    }
                    if !found {
                        conf.checks.push(default_check.clone())
                    }
                }
                self.create_config_file_from_struct(&conf)?;
                Ok(())
            },
            Err(_e) => return Err(anyhow!("could not parse current config file. please try to fix the yaml file or override the current configuration by use the flag `--behavior override`"))
        }
    }
}

/// Get config config application details.
///
/// # Arguments
///
/// * `path` - Config folder path. if is empty default path will be returned.
pub fn get_config_folder(path: &str) -> AnyResult<SettingsConfig> {
    let package_name = env!("CARGO_PKG_NAME");

    let mut config_folder = path.into();
    let mut is_default = false;
    if path.is_empty() {
        match home::home_dir() {
            Some(path) => {
                is_default = true;
                config_folder = format!("{}/.{}", path.display(), package_name);
            }
            None => return Err(anyhow!("could not get directory path")),
        }
    }
    Ok(SettingsConfig {
        path: config_folder.clone(),
        default: is_default,
        config_file_path: format!("{}/config.yaml", config_folder),
    })
}

#[cfg(test)]
mod config {
    use super::*;

    fn get_current_project_path() -> String {
        std::env::current_dir().unwrap().to_str().unwrap().into()
    }
    #[test]
    fn can_get_config_default_folder() {
        let conf = get_config_folder("").unwrap();
        assert_ne!(conf.path, "");
        assert!(conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_get_config_custom_folder() {
        let folder_path = "/custom/folder/path";
        let conf = get_config_folder(folder_path).unwrap();
        assert_eq!(conf.path, folder_path);
        assert!(!conf.default);
        assert_eq!(conf.config_file_path, format!("{}/config.yaml", conf.path));
    }

    #[test]
    fn can_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/src/config.yaml", get_current_project_path()),
        };

        assert!(settings_config.load_config_from_file().is_ok())
    }

    #[test]
    fn cant_load_config_from_file() {
        let settings_config = SettingsConfig {
            path: String::from(""),
            default: false,
            config_file_path: String::from(""),
        };

        assert!(settings_config.load_config_from_file().is_err())
    }

    #[test]
    fn can_write_config_file() {
        let tmp_folder = format!("{}/tmp", get_current_project_path());
        let config_file_path = format!("{}/config.yaml", &tmp_folder);
        if fs::metadata(&config_file_path).is_ok() {
            fs::remove_file(&config_file_path).unwrap();
        }

        let settings_config = SettingsConfig {
            path: format!("{}", tmp_folder),
            default: false,
            config_file_path: config_file_path,
        };

        assert!(settings_config.manage_config_file().is_ok());
        assert!(settings_config.read_config_file().is_ok());
    }

    #[test]
    fn can_load_default_config() {
        let conf = get_config_folder("").unwrap();
        assert!(conf.load_default_config().is_ok())
    }

    #[test]
    fn can_override_existing_config() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/override.yaml", get_current_project_path()),
        };

        // check if we file created successfully
        assert!(fs::File::create(&settings_config.config_file_path)
            .unwrap()
            .write_all("".as_bytes())
            .is_ok());

        // then read the file and make sure that the content is empty
        let file_content = settings_config
            .read_config_file()
            .unwrap_or(format!("error"));

        assert_eq!(file_content, "");

        // create the default configuration
        assert!(settings_config.override_configuration_file().is_ok());
        // make sure that the file is not empty
        assert!(!settings_config.read_config_file().unwrap().is_empty());
    }

    #[test]
    fn can_add_diff_configuration_file() {
        let settings_config = SettingsConfig {
            path: get_current_project_path(),
            default: false,
            config_file_path: format!("{}/tmp/add-diff.yaml", get_current_project_path()),
        };

        let orig_config = settings_config.load_default_config().unwrap();
        let mut config = settings_config.load_default_config().unwrap();

        // creates configuration file with only 1 check (we want to check that we not change existing check and just append new ones)
        config.checks = vec![Check {
            is: String::from("is value"),
            method: Method::Contains,
            enable: true,
            description: String::from("description"),
        }];

        // create configuration file with 1 check
        assert!(settings_config
            .create_config_file_from_struct(&config)
            .is_ok());

        // make sure that the configuration created with 1 check
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len()
        );

        // create the config diff command
        assert!(settings_config.add_diff_configuration_file().is_ok());

        // make sure that the count of check that change equal to existing config + the default application config
        assert_eq!(
            settings_config
                .load_config_from_file()
                .unwrap()
                .checks
                .len(),
            config.checks.len() + orig_config.checks.len()
        );
    }
}
