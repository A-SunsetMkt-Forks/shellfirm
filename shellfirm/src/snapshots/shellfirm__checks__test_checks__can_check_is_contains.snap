---
source: src/checks.rs
expression: "is_contains(\"test\", \"test is valid\")"
---
true
